File: libsrreal/diffpy/Attributes.cpp
/*****************************************************************************
*
* diffpy.srreal     by DANSE Diffraction group
*                   Simon J. L. Billinge
*                   (c) 2009 Trustees of the Columbia University
*                   in the City of New York.  All rights reserved.
*
* File coded by:    Pavol Juhas
*
* See AUTHORS.txt for a list of people who contributed.
* See LICENSE.txt for license information.
*
******************************************************************************
*
* class Attributes - interface for calling setter and getter methods using
*   their string names.
*
* $Id$
*
*****************************************************************************/

#include <sstream>

#include <diffpy/Attributes.hpp>

using namespace std;

namespace diffpy {
namespace attributes {

//////////////////////////////////////////////////////////////////////////////
// class DoubleAttribute
//////////////////////////////////////////////////////////////////////////////

// Public Methods ------------------------------------------------------------

double Attributes::getDoubleAttr(const string& name) const
{
    this->checkAttributeName(name);
    GetDoubleAttrVisitor vg(name);
    this->accept(vg);
    return vg.getValue();
}


void Attributes::setDoubleAttr(const string& name, double value)
{
    this->checkAttributeName(name);
    SetDoubleAttrVisitor vs(name, value);
    this->accept(vs);
}


bool Attributes::hasDoubleAttr(const string& name) const
{
    CountDoubleAttrVisitor vc(name);
    this->accept(vc);
    return vc.count();
}


set<string> Attributes::namesOfDoubleAttributes() const
{
    NamesOfDoubleAttributesVisitor vn;
    this->accept(vn);
    set<string> rv = vn.names();
    return rv;
}

// Private Methods -----------------------------------------------------------

void Attributes::checkAttributeName(const string& name) const
{
    CountDoubleAttrVisitor v(name);
    this->accept(v);
    if (v.count() == 0)
    {
        ostringstream emsg;
        emsg << "Invalid attribute name '" << name << "'.";
        throw invalid_argument(emsg.str());
    }
    else if (v.count() > 1)
    {
        ostringstream emsg;
        emsg << "Duplicate attribute '" << name << "'.";
        throw logic_error(emsg.str());
    }
}

// Visitor Classes -----------------------------------------------------------

// CountDoubleAttrVisitor

Attributes::CountDoubleAttrVisitor::
CountDoubleAttrVisitor(const string& name) :
    mname(name),
    mcount(0)
{ }


void Attributes::CountDoubleAttrVisitor::
visit(const Attributes& a)
{
    mcount += a.mdoubleattrs.count(mname);
}


int Attributes::CountDoubleAttrVisitor::
count() const
{
    return mcount;
}

// GetDoubleAttrVisitor

Attributes::GetDoubleAttrVisitor::
GetDoubleAttrVisitor(const string& name) :
    mname(name)
{ }


void Attributes::GetDoubleAttrVisitor::
visit(const Attributes& a)
{
    DoubleAttributeStorage::const_iterator ai;
    ai = a.mdoubleattrs.find(mname);
    if (ai != a.mdoubleattrs.end())
    {
        mvalue = ai->second->getValue(&a);
    }
}


double Attributes::GetDoubleAttrVisitor::
getValue() const
{
    return mvalue;
}

// SetDoubleAttrVisitor

Attributes::SetDoubleAttrVisitor::
SetDoubleAttrVisitor(const string& name, double value) :
    mname(name),
    mvalue(value)
{ }


void Attributes::SetDoubleAttrVisitor::
visit(Attributes& a)
{
    DoubleAttributeStorage::iterator ai;
    ai = a.mdoubleattrs.find(mname);
    if (ai != a.mdoubleattrs.end())
    {
        ai->second->setValue(&a, mvalue);
    }
}

// NamesOfDoubleAttributesVisitor

void Attributes::NamesOfDoubleAttributesVisitor::
visit(const Attributes& a)
{
    DoubleAttributeStorage::const_iterator ati;
    for (ati = a.mdoubleattrs.begin(); ati != a.mdoubleattrs.end(); ++ati)
    {
        mnames.insert(ati->first);
    }
}


const set<string>& Attributes::NamesOfDoubleAttributesVisitor::
names() const
{
    return mnames;
}


}   // namespace attributes
}   // namespace diffpy

// End of file

File: src/diffpy/srreal/attributes.py
#!/usr/bin/env python
##############################################################################
#
# diffpy.srreal     by DANSE Diffraction group
#                   Simon J. L. Billinge
#                   (c) 2010 The Trustees of Columbia University
#                   in the City of New York.  All rights reserved.
#
# File coded by:    Pavol Juhas
#
# See AUTHORS.txt for a list of people who contributed.
# See LICENSE_DANSE.txt for license information.
#
##############################################################################

"""class Attributes  -- wrapper to C++ class diffpy::Attributes
    A base to PairQuantity and quite a few other classes.
"""

__all__ = ['Attributes']

from diffpy.srreal.srreal_ext import Attributes

# Inject the __getattr__ and __setattr__ methods to the Attributes class

def _getattr(self, name):
    '''Lookup a C++ double attribute if standard Python lookup fails.

    Raise AttributeError if C++ double attribute does not exist.
    '''
    rv = self._getDoubleAttr(name)
    return rv

Attributes.__getattr__ = _getattr


def _setattr(self, name, value):
    '''Assign to C++ double attribute if Python attribute does not exist.
    '''
    try:
        object.__getattribute__(self, name)
    except AttributeError:
        if self._hasDoubleAttr(name):
            self._setDoubleAttr(name, value)
            return
    object.__setattr__(self, name, value)
    return

Attributes.__setattr__ = _setattr

# Helper accessor functions used by the _registerDoubleAttribute method

def _pyattrgetter(name):
    f = lambda obj: object.__getattribute__(obj, name)
    return f

def _pyattrsetter(name):
    f = lambda obj, value: object.__setattr__(obj, name, value)
    return f

# End of file
